codec-http2/src/main/java/io/netty/handler/codec/http2/HttpConversionUtil.java

```diff
@@ -413,9 +415,34 @@ public static Http2Headers toHttp2Headers(HttpHeaders inHeaders, boolean validat
     }
 
     private static CharSequenceMap<AsciiString> toLowercaseMap(List<String> values) {
-        CharSequenceMap<AsciiString> result = new CharSequenceMap<AsciiString>();
+        UnsupportedValueConverter<AsciiString> valueConverter = UnsupportedValueConverter.<AsciiString>instance();
+        CharSequenceMap<AsciiString> result =
+            new CharSequenceMap<AsciiString>(true, valueConverter, values.size());
+
+        // we iterate because the underlying list is probably a linked list
         for (CharSequence value : values) {
-            result.add(AsciiString.of(value).toLowerCase(), EMPTY_STRING);
+            AsciiString lowerCased = AsciiString.of(value).toLowerCase();
+            try {
+                int index = lowerCased.forEachByte(FIND_COMMA);
+                if (index != -1) {
+                    int start = 0;
+                    do {
+                        result.add(lowerCased.subSequence(start, index, false).trim(), EMPTY_STRING);
+                        start = index + 1;
+                    } while (start < lowerCased.length() &&
+                             (index = lowerCased.forEachByte(start, value.length() - start, FIND_COMMA)) != -1);
+                    if (start >= lowerCased.length()) {
```

Comment:
It doesn't seem like this should be special-cased or that all empty strings should be handled (by ignoring them), assuming that's its purpose.

From what I can tell, it is only possible for `start == lowerCased.length()` which is "safe" for `subSequence()`. That would happen for a string like "trailing,comma," which produces an extra empty string. But if we're concerned about that, then it seems ",leading,comma" and "inner,,comma" should also be disallowed.

While RFC 7230 §7 says:
> In any production that uses the list construct, a sender MUST NOT
   generate empty list elements.

It also says:
> For compatibility with legacy list rules, a recipient MUST parse and
   ignore a reasonable number of empty list elements

You may consider using `StringUtil.unescapeCsvFields` for this. It supports quotes, which is unnecessary here, but shouldn't hurt anything other than performance.

Reply:
:( It seems `StringUtil.unescapeCsvFields` doesn't do the trimming, even though it is used by `CombinedHttpHeaders`.